subprojects/worker-processes/src/main/java/org/gradle/process/internal/worker/messaging/WorkerConfigSerializer.java

```
@@ -0,0 +1,115 @@
+        boolean shouldPublishJvmMemoryInfo = decoder.readBoolean();
+        String gradleUserHomeDirPath = decoder.readString();
+        MultiChoiceAddress serverAddress = new MultiChoiceAddressSerializer().read(decoder);
+        final long workerId = decoder.readSmallLong();
+        final String displayName = decoder.readString();
+        Action<? super WorkerProcessContext> workerAction = deserializeWorker(decoder.readBinary(), getClass().getClassLoader());
+
+        return new WorkerConfig(logLevel, shouldPublishJvmMemoryInfo, gradleUserHomeDirPath, serverAddress, workerId, displayName, workerAction);
+    }
+
+    @Override
+    public void write(Encoder encoder, WorkerConfig config) throws IOException {
+        encoder.writeSmallInt(config.getLogLevel().ordinal());
+        encoder.writeBoolean(config.shouldPublishJvmMemoryInfo());
+        encoder.writeString(config.getGradleUserHomeDirPath());
+        new MultiChoiceAddressSerializer().write(encoder, config.getServerAddress());
+        encoder.writeSmallLong(config.getWorkerId());
+        encoder.writeString(config.getDisplayName());
+        encoder.writeBinary(serializeWorker(config.getWorkerAction()));
+    }
+
+    private static Action<? super WorkerProcessContext> deserializeWorker(byte[] serializedWorker, ClassLoader loader) throws IOException {
+        ByteArrayInputStream bais = new ByteArrayInputStream(serializedWorker);
+        ObjectInputStream in = null;
+        try {
+            in = new ClassLoaderObjectInputStream(bais, loader);
+
+            @SuppressWarnings("unchecked")
+            Action<? super WorkerProcessContext> workerAction = (Action<? super WorkerProcessContext>) in.readObject();
+            return workerAction;
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Cannot load worker action's class", e);
+        } finally {
+            if (in != null) {
+                in.close();
```

You should probably handle exceptions on closing the streams within the `finally` blocks rather than throwing those `IOException`s, even if you intend to throw the exception from the body of the `try`s out of these methods.  Looks like we typically don't use utils like `IOUtils.closeQuietly()` or anything but just include another nested try-catch.

---

What exactly would you do in the case of a failed close? What is there to do with that exception other than throw it?

---

Log it and continue running?  That sort of failure doesn't seem like it should have to potential to be fatal.

---

Added a log. I searched for similar occurrences (`finally \{\s*if\s*\(.*\!=.*null.*\)\s*\{\s*`) there seems to be few different things happening. 
* Throw the IOException (same as this code)
* Throw as un unchecked exception
* Ignore the exception without logging

No existing cases logged the warning, though I updated this code to log at the `DEBUG` level since that seems like the best solution.